cmd/kube-apiserver/app/options/validation.go

```diff
@@ -102,13 +106,33 @@ func validateServiceNodePort(options Extra) []error {
 	return errs
 }
 
+func validatePublicIPServiceClusterIPRangeIPFamilies(extra Extra, generic genericoptions.ServerRunOptions) []error {
+	// The "kubernetes.default" Service is SingleStack based on the configured ServiceIPRange.
+	// If the bootstrap controller reconcile the kubernetes.default Service and Endpoints, it must
+	// guarantee that the Service ClusterIPRepairConfig and the associated Endpoints have the same IP family, or
+	// it will not work for clients because of the IP family mismatch.
+	// TODO: revisit for dual-stack https://github.com/kubernetes/enhancements/issues/2438
+	if reconcilers.Type(extra.EndpointReconcilerType) != reconcilers.NoneEndpointReconcilerType {
+		serviceIPRange, _, err := controlplaneapiserver.ServiceIPRange(extra.PrimaryServiceClusterIPRange)
+		if err != nil {
+			return []error{fmt.Errorf("error determining service IP ranges: %w", err)}
+		}
+		if netutils.IsIPv4CIDR(&serviceIPRange) != netutils.IsIPv4(generic.AdvertiseAddress) {
```

Comment:
is this called after the AdvertiseAddress is obtained here?

https://github.com/kubernetes/kubernetes/blob/8d0e46a5ec38faed3e7277710103b37fa9140378/staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go#L133-L149

Reply:
yes, it's called in `Complete()`, i.e. before this validation.